design/01-proc-macro.md

```diff
@@ -0,0 +1,290 @@
+- The procedural macro will be triggered by one of three Cairo expressions
+    - Macro call, e.g. `xyz!`
+    - Macro attribute, e.g. `#[xyz]`
+    - Macro derive, e.g. `#[derive(Xyz)]`
+- A Rust crate exposing an API for writing procedural macros for Cairo will be published for programmers. This crate will be automatically added to generated `Cargo.toml` file.
+- The procedural macro implementation will be a Rust function, accepting a `TokenStream` (described in detail in following sections) on input and returning the expansion result as an output.
+- The procedural macro implementation will return a `ProcMacroResult` defined in the helper library. The enum will contain one of three values:
+    - `ProcMacroResult::Leave` - procedural macro does not take any action
+    - `ProcMacroResult::Replace({ TokenStream, AuxData })` - procedural macro has generated code that replaces input
+    - `ProcMacroResult::Remove` - original code should be removed
+    - `ProcMacroResult::Bail(ProcMacroDiagnostics)` - procedural macro exited with an error
+- Alongside the new TokenStream, a procedural macro can emit auxiliary data, encoded as an arbitrary JSON.
+- The appropriate procedural macros will be executed based on the call in Cairo code by the new Cairo compiler internal `ProcMacroHost` plugin. This plugin will be called on each AST node and will decide if analyzed fragment requires code generation powered by an external plugin.
+
+**Example:**
+
+The procedural macro source code:
+
+```rust
+pub fn generate_code(token_stream: TokenStream) -> ProcMacroResult {
+	// Implementation skipped for brevity.
+}
+```
+
+The communication between Scarb and the procedural macros will have to be implemented in such a manner, that:
+
+- Will be forward and backward compatible regardless of changes in the Cairo compiler internals (including the Cairo AST).
+- Can be established over the FFI barrier.
+
+In order to meet the specified requirements, instead of syntax tree nodes, we can rather pass plain Cairo tokens which will be parsed by the procedural macro itself.
+
+A token can be defined as a string with one of three kinds assigned to it during parsing. The token kinds are: identifier, punctuation symbol and literal. Kinds of different tokens can be represented as an enum called TokenTree. An iterator over TokenTrees is called a TokenStream. A TokenStream is a representation of a Cairo code.
+
+Implementation of the communication protocol will require some adjustments in the Cairo language parser, so that it can be used for TokenStreams. Namely, it should be possible to use TokenStreams as input instead of Cairo files, skipping the lexical analysis required for Cairo files (as TokenStream can already be analyzed on the compiler side).
+
+## **Drawbacks**
+
+- Since shared libraries are loaded into the process memory space, some errors in the execution might result in hard to debug segmentation faults.
```

Comment:
I've been thinking of extracing Scarb's compiler machinery into a separate binary, that would be run as separate process. That level of isolation would at least prevent memory corruption from killing entire Scarb execution, so `scarb build` will at least be able to recover with good error message.

Reply:
I will note this in "future possibilities" section if you are ok with that.

Reply:
That would be great!